README.md

```diff
@@ -1,26 +1,102 @@
+have a `value` prop, it's an uncontrolled component. Uncontrolled components can optionally have a `defaultValue` prop
+to set its initial value. These two props (`value` and `defaultValue`) are _mutually exclusive_: a component is either
+controlled or uncontrolled, but it cannot be both.
+
+### Uncontrolled component
+In uncontrolled components, the state of the component's value is kept in the DOM itself — in other words, the form
+element in question (e.g. an `<input>`) has its _own internal state_. To retrieve that value, we would need direct
+access to the DOM component that holds the value, _or_ we'd have to add an `onChange` handler.
+
+To set an initial value for the element, we'd use the `defaultValue` prop. We can't use the `value` prop for this: we're
+not using state to explicitly store its value, so the component would never update its value anymore (since we're
+rendering the same thing).
+
+### Controlled component
+In controlled components, we explicitly set the value of a component, and update that value in response to any changes
+the user makes to the value of that component. That might sound a little wonky, but when you see the code, it'll become
+much clearer:
 
+```js
+class ControlledInput extends React.Component {
+  constructor() {
+    super();
+    
+    this.handleChange = this.handleChange.bind(this);
+    
+    this.state = {
+      value: '',
+    };
+  }
+  
+  handleChange(ev) {
+    this.setState({
+      value: ev.target.value,
+    });
+  }
+
+  render() {
+    return (
+      <input type="text" value={this.state.value} onChange={this.handleChange} />
+    );
+  }
+}
+```
+
+As you can see, we can easily define the initial value by setting the initial `value` property on the state to whatever
+we want. 
+
+Using a controlled component is the preferred way to do things in React — it allows us to keep _all_ component state in
+the React state, instead of relying on the DOM to retrieve the element's value through its internal state. Whenever our
+state changes, the component re-renders, rendering the input with the new updated value. If we don't update the state,
+our input wouldn't update when the user would type. In other words, we need to update our input's state _programatically_.
+
+It might seem a little counterintuitive that we need to be so verbose, but this actually opens the door to additional
+functionality. For example, let's say we want to write an input that only takes in a number (let's pretend there is no
+`<input type="number">`). We can now validate the data the user enters _before_ we set it on the state, allowing us to
+block any invalid values. If the input is invalid, we simply don't set the state, not updating the input. We could
+optionally set another state property (for example, `isInvalidNumber`). Using that state property, we can show an error
+in our component to indicate that the user tried to enter an invalid value.
```

Comment:
Very good illustration of your point about the added power of a controlled component.